File: openmodels/adapter.py
from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Any


class AdapterError(Exception):
    pass


@dataclass(frozen=True)
class GeneratedFile:
    path: str
    content: str


class BackendAdapter(ABC):
    key: str
    description: str
    default_filename: str

    @abstractmethod
    def generate_files(
        self,
        canonical_model: dict[str, Any],
        filename: str | None = None,
        options: dict[str, Any] | None = None,
    ) -> list[GeneratedFile]:
        raise NotImplementedError

File: rust/openmodels-rs/src/adapter.rs
use crate::error::Result;
use crate::model::{CanonicalModel, JsonObject};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GeneratedFile {
    pub path: String,
    pub content: String,
}

pub trait BackendAdapter: Sync {
    fn key(&self) -> &'static str;
    fn description(&self) -> &'static str;
    fn default_filename(&self) -> &'static str;
    fn generate_files(
        &self,
        canonical_model: &CanonicalModel,
        filename: Option<&str>,
        options: Option<&JsonObject>,
    ) -> Result<Vec<GeneratedFile>>;
}
